src/bsoncxx/json.hpp

```diff
@@ -34,18 +34,42 @@ BSONCXX_INLINE_NAMESPACE_BEGIN
 extern const bool k_silence_doxygen;
 
 ///
-/// Converts a BSON document to a JSON string.
+/// Converts a BSON document to a JSON string, in legacy extended format.
 ///
 /// @param view
 ///   A valid BSON document.
 ///
 /// @throws bsoncxx::exception with error details if the conversion failed.
 ///
-/// @returns A JSON string.
+/// @returns A legacy extended JSON string.
 ///
 BSONCXX_API std::string BSONCXX_CALL to_json(document::view view);
 
 ///
+/// Converts a BSON document to a JSON string, in canonical extended format.
+///
+/// @param view
+///   A valid BSON document.
+///
+/// @throws bsoncxx::exception with error details if the conversion failed.
+///
+/// @returns A canonical extended JSON string.
+///
+BSONCXX_API std::string BSONCXX_CALL to_canonical_extended_json(document::view view);
```

Comment:
Not that the modes are likely to change very much in the future, but I would generally opt for a single method with public enumeration:

```
BSONCXX_API std::string BSONCXX_CALL to_extended_json(document::view view, ExtendedJsonMode mode);
```

Is there an obvious reason to prefer three explicit methods that I'm missing? Seems like more symbols to maintain, and not very extensible from an ABI perspective.

Reply:
I think this is the most critical part of the review, since it is the user facing API. Lets sort this one out. Arguments for three methods, rather than public enum and one method?

Reply:
We aren't currently enforcing ABI stability. So it would still be possible to add a new enum parameter that took a defaulted value. That would still get you API stability. Had we declared ABI stability, I'd agree with you. But since we haven't, I don't think we should use it to decide API design decisions.

Reply:
I might also add that for `to_json` we are returning the "JSON" representation, with `to_extended_json` we are returning the "Extended JSON" representation, those are the two things we should provide API for.  Returning "relaxed" or "canonical" Extended JSON is an _optional_ variant of Extended JSON, which is why I see it as an enum passed to a single function.  Originally we had a "strict" mode for Extended JSON as well, so you can see there is potential for growth in this enum.   So thats my stylistic argument, beyond my "less exported symbols is better" argument :smile:

Reply:
Also, from the spec:
```
A generator MUST allow users to produce strings in either the Canonical Extended JSON format or the Relaxed Extended JSON format. If generators provide a default format, the default SHOULD be the Relaxed Extended JSON format.
```